### DISCUSSION TITLE
Scalable Keyboard

### DISCUSSION THREAD

**Discussion by lageist** [UNC-Project-Open-AAC/OS-DPI#133] [Ideas]
Design aim: Create a scalable keyboard that can be efficiently adjusted based on access method. Example will be set to full screen mode for eye gaze access and scaled down to about a quarter of the screen for trackball access. 

Source: Community member input

Sketches:
![keybd_100%](https://user-images.githubusercontent.com/91340176/210388372-a617cdac-17d8-4110-abba-bf6ddbe90b99.png)
![keybd_25%](https://user-images.githubusercontent.com/91340176/210388430-2778e100-3155-45b4-a32e-f39669763f15.png)

**Comment by gbishop** [comment]
That should be straightforward. I think we should scale the fonts down with
the keyboard. That isn't currently supported but would be easy with the new
Customize component.

The easiest approach would be a tab control (with hidden tabs) representing
the keyboard. The tabs are identical except the second has stacks and gaps
to force the keyboard into the lower corner. On the customize component
you'd pick out the 2nd tab and set the font size smaller.

An alternative would be to alter the grid css so the font size is
proportional (somehow) to the size of the button. I think that is doable
and might be generally better.

gb

**Reply by lageist** [reply]
Thanks @gbishop. I'll try using the above - much simpler approach. I was plugging away exploring using stacks and scaling, which works, mostly, but is more complicated than it needs to be.    
![scale_25%](https://user-images.githubusercontent.com/91340176/210406727-093fed33-1455-4b46-86ad-b536297e251f.png)
![scale_full](https://user-images.githubusercontent.com/91340176/210406745-f3fecbd2-ffb0-40e2-9fed-f1f7f5297a7d.png)

**Comment by gbishop** [comment]
With the new Copy/Paste functions you should be able to create the entire
tab then paste it into the other. No need to create it twice.

**Reply by lageist** [reply]
Yes, copy/paste was a huge timesaver here! The current version of this work in progress is [here](https://github.com/UNC-Project-Open-AAC/OS-DPI/blob/6f3ef4acd70a241146af81736be4e6a61c59beb2/Designs%20In%20Progress/LG_Explore/scalable_keyboard_010323.osdpi). Welcome any suggestions.

**Comment by lageist** [comment]
Current version of this project is attached (testing if sharing the .osdpi as .zip file in discussion works for community sharing).
[scalable_keyboard_010423.osdpi.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/10347945/scalable_keyboard_010423.osdpi.zip)

**Comment by sofia-bg** [comment]
Lori this looks great! Since the user this was developed with in mind alternates between track ball and eye-gaze, the pointer dwell method really does work for both access methods right? So that the grid is changing when he indicates he is using a different method,, but on our end we don't need to actually specify that either are different because they are essentially both point and dwell, correct?

**Reply by lageist** [reply]
Yes, I think that's right.

**Comment by aetkie** [comment]
Revisiting this idea with 2 added components. 
- a pop out or toggle on/off grid of navigation buttons (pictured in red)
- morph forms function (not pictured)
<img width="1088" alt="scalable_kybd_v02_sketch_AE" src="https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/113473933/80392bdb-e7ae-4ff6-8361-ce332ddb3e25">

**Comment by gbishop** [comment]
Looks great!

I hallucinate that T9 plus morphs could  make an interesting small keyboard.

**Comment by sofia-bg** [comment]
Here is a UI that auto-scales from small to full when head mouse is turned on. Then back to small when head mouse turned off. t turns headmouse on; T turns headmouse off. s opens headmouse settings. 

[Auto-scale keyboard](https://unc-project-open-aac.github.io/OS-DPI/?fetch=https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14858664/autoscalable_kybd_4_3_24_SBG.zip)

@lageist

**Comment by gbishop** [comment]
I get multiple characters entered when I use the mouse. The issue appears
to be that you have a Pointer down handler in both the Pointer dwell method
and in the Mouse method.

Do you need both of them active?

Also, I have learned that using Pointer up is superior to Pointer down in
just about every circumstance.

**Reply by sofia-bg** [reply]
Oh interesting because I intermittently have this problem across my layouts and haven't been able to figure out why. But the weird thing is is that all I did was use the layout that Lori made here and did not make any changes to the mouse method and her's works just fine:
[scalable_kybd_v01_010423_lg.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14872619/scalable_kybd_v01_010423_lg.zip)

Can poke around more to see if something happened in the uploading/exporting that maybe changed something between our two versions. 

Also I have only ever changed the pointer handler for when I want it to trigger on enter, otherwise the standard for mouse is set to down so I have always just used that. Will play around

**Reply by sofia-bg** [reply]
never mind seeing the problem is that i turned on pointer dwell... don't i need pointer dwell on for the headmouse to work? but the user we are imagining doesn't use mouse to dwell (or maybe he does, do you know @lageist ?)

**Reply by sofia-bg** [reply]
Ok updated version [here](https://github.com/UNC-Project-Open-AAC/OS-DPI/assets/92817124/a99b2e71-52d4-47b4-980d-42acf2415de1). For me it is not doubling the letters on mouse, I changed the dwell to activate on "enter."  

I also replaced the stacks with gaps so that the background is white and the word "empty" is gone.

**Reply by aetkie** [reply]
@sofia-bg can you post the zip for the updated version? I can't open it using the link for some reason

**Reply by sofia-bg** [reply]
[autoscalable_kybd_4_4_24_SBG.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/14920110/autoscalable_kybd_4_4_24_SBG.zip)
@aetkie

**Comment by gbishop** [comment]
The head tracker only moves the mouse. Your subject will need some way to
click.

**Comment by aetkie** [comment]
Latest iteration of this scalable keyboard. I have the color scheme as close as I can get it to the model we were given by the intended user. I have not figured out how to pop out the navigation tools so that button is not in this version.

[autoscalable_kybd_041724_AE.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15012768/autoscalable_kybd_041724_AE.zip)

**Comment by gbishop** [comment]
Nice. You have zipped the .osdpi file which means it won't work when people
try to directly import it. The preferred way is to change the suffix from
.osdpi to .zip. Maybe I can fix that in the app? I think I could unzip,
then look to see if the content is an .osdpi file and process that... I'll
add a bug report.

This also shows a bug on my chromebook. If I mouse over the Enter button
from the right, it doesn't highlight. What's up with that?

**Reply by aetkie** [reply]
I still get confused with those steps. I exported the design. Compressed it. and then took 'osdpi' out of the file name so that its just .zip

**Reply by gbishop** [reply]
I'll look into improving this. The complexity should be in the code, not in your workflow.

**Reply by sofia-bg** [reply]
@aetkie what i do is when I export and it prompts me to save/name i just change the ending then from .osdpi to .zip and that seems to work

**Comment by sofia-bg** [comment]
@aetkie I made this first stab at the navigation expanding to the right. It only is functional right now in the "small" keyboard. You'll see I shifted the keyboard more towards the center so that when the navigation pops out it doesn't shift the keyboard. Just a style preference on my part, could have easily left the keyboard in the same place and have it shift when the navigation grid appears. 

I tried the modal pop-up but to have it retain the keyboard was going to be a lot more copy-paste/redundancy. If you want a modal popup that only shows the navigation buttons that would be easy to do. Let me know if you want a model if that to play with.

[autoscalable_kybd_4_18_24_expand.zip](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/15028128/autoscalable_kybd_4_18_24_expand.zip)

**Reply by aetkie** [reply]
This is neat. I like how you shifted the KB towards the center. I'd be interested to see how you would make the modal pop-up work. I was trying to make that happen but I can't figure out the correct arrangement in the layout. Where do I put it? The option to Add --> Modal dialog doesn't always show up if certain components of the layout are selected. Does that make sense?

**Comment by gbishop** [comment]
If I remember correctly, you can only add a modal to the Page.

**Comment by aetkie** [comment]
One iteration of this scalable KB is viewable [here in Show & Tell.](https://github.com/UNC-Project-Open-AAC/OS-DPI/discussions/250)
